### ISSUE TITLE
Define own domain at installation e.g. allowing full installation for MicroK8s and other DEV plattforms.

### ISSUE THREAD

**Issue by sergiohinojosa** [keptn/keptn#1482]
Hi,

this feature request is to give the possibility to dev plattforms not attached to a LoadBalancer (meaning in status pending) to install all keptn usecases. This can be achieved by passing the domain as a parameter. For example:

```bash
keptn install --platform=kubernetes --istio-install-option=Overwrite --gateway=LoadBalancer --domain=this-is-my-ip-nip.io
```

Before the installation the user has to make sure the domain is reachable. I create the `istio-system` namespace and add map the NGINX ingress listening to port 80 and 443 to route all incoming traffic to the ingress-gateway. Meaning the istio ingress-gateway will work via NodePort and also via https.

``` bash
#Hardcoded Hack for getting the public ip of the singlenode cluster. Tested on microk8s 1.15 Ubuntu 18.
#Change dots for slash (prettier) #and add nip.io domain for subdomain routings. 
#FILE: setupIstio.sh

#Domain used for routing to keptn services
if [[ "$GATEWAY_TYPE" == "LoadBalancer" ]]; then
  printf "About to configure Keptn for Microk8s"
  export PUBLIC_IP=$(curl -s ifconfig.me) 
  PUBLIC_IP_AS_DOM=$(echo $PUBLIC_IP | sed 's~\.~-~g')
  export DOMAIN="${PUBLIC_IP_AS_DOM}.nip.io"
  export INGRESS_HOST=$DOMAIN
```
I completely ommit the validation of the` status.loadBalancer.ingress[0].hostname` or `ip` and set the `DOMAIN `and `INGRESS_HOST `manually. The ConfigMap and the Keptn API VirtualService will be set properly. This way you can spin single node kubernetes instances reachable from the cloud in seconds with a full Keptn installation on microk8s.

**Comment by christian-kreuzberger-dtx** [comment]
Hi! Thanks for the suggestion.

We think Keptn should go even further and provide the `--domain` option in the installer for all platforms, and let the user choose whether they want to use nip.io, xip.io or any other wildcard DNS resolver.

> I completely ommit the validation of the status.loadBalancer.ingress[0].hostname or ip and set the DOMAIN and INGRESS_HOST manually.

That's fine, if you take care of this by yourself. However, I think we might run into a timing issue if we don't wait for the ingress gateway to be ready (AWS and GCP sometimes take their time when setting this up).

However, I think your script will not work for all network configurations. 
Your Public IP is not necessary the one that you want to use for a local development machine. 

I just tried this:
```console
export PUBLIC_IP=$(curl -s ifconfig.me) 
PUBLIC_IP_AS_DOM=$(echo $PUBLIC_IP | sed 's~\.~-~g')
export DOMAIN="${PUBLIC_IP_AS_DOM}.nip.io"
export INGRESS_HOST=$DOMAIN
```
and the domain I get is basically my public IP, but I'm sitting behind a firewall and have NAT active, so via this domain/IP I will not be able to reach my cluster.

Anyway, we will investigate this option further in the near future.

**Comment by sergiohinojosa** [comment]
Hi Christian,

yes, there is no logic nor validation in the script whatsoever. I just use that hardcoded option to get the public ip of an elastic cloud instance in aws. I'll build an image for someone in my team that really wants to use keptn 6.0 in microk8s with vagrant in his own  Laptop. For that workaround I'll save the domain in a ConfigMap and retrieve it at installation. I still believe the best option is to pass the `--domain` as a parameter.
This option is just to be considered for allowing developers test keptn by giving them more flexibility in using kubernetes developer instances and should be used with caution (or prerequisites). 

Best
Sergio

**Comment by sergiohinojosa** [comment]
Just a side node:

I solve the issue momentarely with the following woraround:

Pass the domain into the default namespace:

``` bash
kubectl create configmap keptn-domain --from-literal=domain=$DOMAIN
```
Then read it from the keptn installer pod with an own installer:

````bash
# Excerpt from the setupIstio.sh file

print_info "Check if there is a custom domain to override"
  CUSTOM_DOMAIN=$(kubectl get configmap keptn-domain -o=jsonpath='{.data.domain}')
  if [ $CUSTOM_DOMAIN ]; then
      print_info "Custom domain found $CUSTOM_DOMAIN\n Warning, there will be no validation of the domain nor istio-ingressgateway\n"
      export DOMAIN=$CUSTOM_DOMAIN
      export INGRESS_HOST=$DOMAIN
  else
```

**Comment by christian-kreuzberger-dtx** [comment]
Thanks, I'll take a look at this approach.

I also hit a similar issue today, and I believe that using `curl ifconfig.me` is a viable option for determining the external IP address of a cluster.

However, we need to make sure that we don't just focus on this use-case, but on all use-cases that we cover with Keptn, and one of them is a local developer installation ;)

I'll keep you updated with our findings.

**Comment by sergiohinojosa** [comment]
Hi Christian,
Ok, great. Just as an FYI, An SE college in DACH uses my approach and as domain he pases his home network IP 192.168... as a magic nip.io domain. With a vagrant file he spins his VirtualBox with a Keptn full installation on microk8s. So a nice and generic approach would be to pass the domain as a variable like --domain=my-domain.com. If you want to validate it and it can't be resolved, then you can do a curl to ifconfig.me as a fallback option. 
But just by adding the option of --domain=domain.com it gives more flexibility not only for dev instances but for all operators of K8 clusters. This will allow any keptn installation to be used and exposed via an nginx ingress that has a different IP and a different domain (with maybe a valid certificate). The istio ingressgateway traffic can be routed to this ingress. That's actually what I do. I also use that ingress to manage my certificates and expose the kubernetes API (so I can talk to my dev instance from outside). For instance, I was talking to a customer from the Keptn community and he is routing the istio ingressgateway traffic to an NGINX Ingress already exposed and with a valid certificate.

**Comment by christian-kreuzberger-dtx** [comment]
Thanks for clarification @sergiohinojosa ! 

Going forward, I suggest the following implementations within Keptn:

1. `keptn install --domain=some.domain.com` (where `some.domain.com` could as well be `127.0.0.1.nip.io`) - if domain is provided, we skip any step within the installer to verify the IP address and whether the service is available from the outside - see implementation:
  https://github.com/keptn/keptn/blob/a43adc07d6f9779430347685d5fb859cb6639733/installer/scripts/common/utils.sh#L167-L171

2. For gateway type `NodePort` we need to differentiate whether the user wants the LAN IP of the cluster (e.g., when using MicroK8s or Minikube locally), or the public IP of the system it runs on (think about an EC2 instance or a GCE VM). At the moment, we only retrieve the private cluster IP - see implementation:
  https://github.com/keptn/keptn/blob/a43adc07d6f9779430347685d5fb859cb6639733/installer/scripts/common/utils.sh#L193-L198

3. I believe that we should introduce a third gateway type `NodePortPublicIP`, where we use `curl ifconfig.me` to determine the public IP of the cluster, as suggested by you.

In addition, I think a flag like `--domain-suffix=nip.io` or `--domain-suffix=xip.io` (or any other dynamic resolver) would be beneficial as well.

@agrimmer @grabnerandi what's your take on my three proposed changes?

**Comment by christian-kreuzberger-dtx** [comment]
An experimental implementation with the `--domain` flag is available within PR #1674.

**Comment by agrimmer** [comment]
I like the new option `--domain` in the install command!
